test/integration/device-base.js

```diff
@@ -79,4 +79,27 @@ describe('device-base', function desc() {
 			});
 		});
 	});
+
+
+	describe('openNativeUsbDevice()', () => {
+		it('it can open a native USB device handle', async () => {
+			if (devs.length < 1) {
+				throw new Error('This test requires a device');
+			}
+			const dev1 = devs[0];
+			dev1.open();
+			const id1 = dev1.id;
+			dev1.close();
+
+			const nativeUsbDevice = dev1._dev;
+
+			const dev2 = openNativeUsbDevice(nativeUsbDevice);
+			dev2.open();
+			const id2 = dev2.id;
+			dev2.close();
+
+			expect(id1).to.equal(id2);
```

Comment:
ah, ok. i figured out how to actually run these. i'll update the docs so that's clearer but in general i'm thinking we can start moving things over to / preferring the `e2e` tests.

to run these, connect a wifi and cellular device to usb, then run: `RUN_INTEGRATION_TESTS=1 npm test` (see [here](https://github.com/particle-iot/particle-usb/blob/f303fd28220bff1aa522ffd0fe268e2082e92e49/test/support/index.js#L34)) 

doing that, i see an error: `TypeError: dev2.open is not a function`

then i noticed that the calls to `.open()`, `.close()`, and `openNativeUsbDevice()` aren't using the `await` keyword:

```suggestion
			if (devs.length < 1) {
				throw new Error('This test requires a device');
			}
			const dev1 = devs[0];
			await dev1.open();
			const id1 = dev1.id;
			await dev1.close();

			const nativeUsbDevice = dev1._dev;

			const dev2 = await openNativeUsbDevice(nativeUsbDevice);
			await dev2.open();
			const id2 = dev2.id;
			await dev2.close();

			expect(id1).to.equal(id2);
```

with that in place, i see a different error: `TypeError: Cannot read properties of undefined (reading 'idVendor')` which is the same one i got over in https://github.com/particle-iot/particle-usb/pull/75/files#r998788868

Reply:
@rickkas7 @sergeuz 

per https://github.com/particle-iot/particle-usb/pull/75#discussion_r1001217762 i think the "fix" here comes down to:

1. get proper native usb device: `const nativeUsbDevice = dev1._dev._dev;`
2. remove the `dev2.open();` call as it is redundant
3. add in the `await` keywords

```suggestion
			if (devs.length < 1) {
				throw new Error('This test requires a device');
			}
			const dev1 = devs[0];
			await dev1.open();
			const id1 = dev1.id;
			await dev1.close();

			const nativeUsbDevice = dev1._dev._dev;

			const dev2 = await openNativeUsbDevice(nativeUsbDevice);
			const id2 = dev2.id;
			await dev2.close();

			expect(id1).to.equal(id2);
```

Reply:
Yeah, the constructor of `DeviceBase` doesn't take a native device handle directly, it expects it to be wrapped into a low-level `UsbDevice` instance which provides a common interface for interacting with the device in Node and browsers. The suggested change looks good to me. There seems to be a public API for accessing the internal handle, although it's a bit awkward: `dev.usbDevice.internalObject`.

Reply:
ah, ok. i updated that suggested change to use the `device.usbDevice.internalObject` path. 

@rickkas7 for consistency, it's probably worth updating the `e2e` test as well 👉 https://github.com/particle-iot/particle-usb/blob/feature/open-native-usb/test/e2e/node.e2e.js#L103

Reply:
@rickkas7 ☝️ in case it got lost in the shuffle, we still need to apply these changes as currently that integration test will always fail

Reply:
Oops, missed that. I made the change and it passes npm run test:e2e now.

Reply:
great, thanks! FYI these are the integration tests. you run them with: `RUN_INTEGRATION_TESTS=1 npm test` (see https://github.com/particle-iot/particle-usb/pull/75#discussion_r999640280)

Reply:
Hmm... I'm getting an unexpected error running the e2e test:

```
rickk@iMac particle-usb % RUN_INTEGRATION_TESTS=1 npm test

> particle-usb@2.1.3 test
> mocha test test/integration './src/**/*.test.js'



  device
    Device
      with multiple devices
        ✔ open, reset, close (473ms)

  cellular-device
    CellularDevice
      getIccid()
        ✔ gets ICCID of the active SIM card

  device-base
    openDeviceById()
      ✔ does not affect devices which are already open
    DeviceBase
      sendControlRequest()
        1) can process multiple requests
      sendControlRequest()
        ✔ can send a request and receive a reply larger than 4096 bytes
    openNativeUsbDevice()
      ✔ it can open a native USB device handle

  device-quirks
    controlOutTransfersRequireDataStage
      ✔ disconnectFromCloud() does not timeout

  CellularDevice
    ✔ provides getIccid() for Particle boron

  CloudDevice
    ✔ provides CloudDevice methods in cloud-device.js (which ALL Particle Devices inherit from)

  device-base
    getDevices()
      ✔ enumerates only Particle USB devices
      ✔ includes devices in the DFU mode by default
      ✔ can optionally exclude devices in the DFU mode
      ✔ can filter detected devices by type
      ✔ matches device types in a case-insensitive manner
      ✔ ignores invalid device types
    openDeviceById()
      ✔ opens a device by ID
      ✔ fails if the device cannot be found
      ✔ ignores non-Particle devices with a matching serial number
      ✔ matches serial numbers in a case-insensitive manner
    openNativeUsbDevice()
      ✔ opens a native usb device
      ✔ opens throws an exception on invalid device
    DeviceBase
      with single device
        open()
          ✔ opens the device
          ✔ initializes the device ID and firmware version properties of the device object
          ✔ filters out non-printable ASCII characters from the device ID string
          ✔ converts the device ID string to lower case
          ✔ does not require the USB device to support the firmware version request
          ✔ resets all pending requests after opening the USB device
          ✔ fails if the USB device was detached from the host
          ✔ fails if the device is already open
        close()
          ✔ closes the device
          ✔ succeeds if the device is already closed
          ✔ can cancel pending requests before closing the device
          ✔ cancels pending requests when the timeout is reached
          ✔ resets active requests when the timeout is reached
        sendControlRequest()
          ✔ can send a request without payload data
          ✔ can send a request with payload data
          ✔ can send request data in multiple chunks
          ✔ can receive a reply without payload data
          ✔ can receive a reply with payload data
          ✔ can receive reply data in multiple chunks
          ✔ polls the USB device until it allocates a buffer for the request data
          ✔ polls the USB device until it completes processing of a request
          ✔ resolves to an object containing the result code
          ✔ resolves to an object containing the reply data
          ✔ sends requests to the USB device concurrently
          ✔ limits the maximum number of concurrent requests
          ✔ times out after a specified amount of time
          ✔ resets an active request when it times out
          ✔ converts the reply data to a string if the request data is a string
          ✔ fails if the device is closed or being closed
          ✔ fails if the request type is not within the range of valid values
          ✔ fails if the request data is too large
      with multiple devices
        ✔ open and close

  Device
    ✔ provides getSerialNumber()
    ✔ provides enterListeningMode()
    ✔ implements enterListeningMode() in a way that will catch errors if the device does not support this request type
    ✔ provides sendProtobufRequest() (this is the low-level API that all higher level APIs use)
    ✔ implements sendProtobufRequest() in a way that throws RequestError when control request is not OK
    ✔ implements sendProtobufRequest() in a way that creates a valid decoded protobuf message even if sendControlRequest does not return valid .data field

  Gen3Device
    ✔ provides setSetupDone() function for argon
    ✔ provides setSetupDone() function for boron

  NetworkDevice (all deprecated methods)
    ✔ provides getNetworkStatus
    ✔ provides getNetworkConfig
    ✔ provides setNetworkConfig

  Public interface of npm module
    ✔ exports expected objects and functions

  setDevicePrototype(), a helper function that sets prototype and inheritance hierarchy based on hardware platform
    ✔ returns a Device instance for all platforms

  WifiDeviceLegacy
    ✔ provides deprecated wifi functions for core, photon, and p1 devices

  WifiDevice
    expected interface for gen3+ wifi devices
      ✔ includes scanWifiNetworks Function for argon platform
      ✔ includes joinNewWifiNetwork Function for argon platform
      ✔ includes clearWifiNetworks Function for argon platform
      ✔ includes scanWifiNetworks Function for asom platform
      ✔ includes joinNewWifiNetwork Function for asom platform
      ✔ includes clearWifiNetworks Function for asom platform
      ✔ includes scanWifiNetworks Function for p2 platform
      ✔ includes joinNewWifiNetwork Function for p2 platform
      ✔ includes clearWifiNetworks Function for p2 platform
    behaviors
      scanWifiNetworks()
        ✔ returns empty when no networks are returned
        ✔ returns empty when pass:false (failure for known/normal reasons like TimeoutError or wrong reply message)
        ✔ returns valid networks with strings for security fields rather than integers
        ✔ converts security integer values to meaningful values based on existing protobuf enum
        ✔ sets ssid to null if Device OS returns with undefined
      joinNewWifiNetwork()
        ✔ Sends wifi.JoinNewNetworkRequest protobuf message with correct data
        - Can Join open Wifi network without security/password
      clearWifiNetworks()
        ✔ Sends wifi.ClearKnownNetworksRequest protobuf message
      _sendAndHandleProtobufRequest
        ✔ Calls sendProtobufRequest with correct arguments
        ✔ Does not pass when sendProtobufRequest returns undefined
        ✔ Does not pass when sendProtobufRequest throws TimeoutError with the default timeout
        ✔ Does not pass when sendProtobufRequest throws TimeoutError with an override timeout message
        ✔ Throws errors other than TimeoutError


  87 passing (7s)
  1 pending
  1 failing

  1) device-base
       DeviceBase
         sendControlRequest()
           can process multiple requests:
     MemoryError: Memory allocation error
      at /Users/rickk/Documents/src/Particle/particle-usb/src/device-base.js:543:12


```

Reply:
@rickkas7 that one passes for me here so i'm thinking it's just a flaky test (not uncommon in these sorts of tests). I say let's create a story in Shortcut for it and press on 😁 👍